<a href="https://colab.research.google.com/github/wangbingo/LVS/blob/master/201909140200__attetion_train_12e__about_8H.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# init system 

In [15]:
!ls

drive  mydrive	sample_data


In [0]:
!rm -rf mmdetection/

In [16]:
!pip3 install torch==1.2.0+cu92 torchvision==0.4.0+cu92 -f https://download.pytorch.org/whl/torch_stable.html

Looking in links: https://download.pytorch.org/whl/torch_stable.html
     |████████████████████████████████| 663.1MB 36kB/s 
     |████████████████████████████████| 8.8MB 43.2MB/s 
  Found existing installation: torch 1.1.0
    Uninstalling torch-1.1.0:
      Successfully uninstalled torch-1.1.0
  Found existing installation: torchvision 0.3.0
    Uninstalling torchvision-0.3.0:
      Successfully uninstalled torchvision-0.3.0


In [17]:
!git clone https://github.com/wangbingo/mmdetection.git

Cloning into 'mmdetection'...
remote: Enumerating objects: 70, done.
remote: Counting objects: 100% (70/70), done.
remote: Compressing objects: 100% (47/47), done.
remote: Total 5233 (delta 45), reused 47 (delta 23), pack-reused 5163
Receiving objects: 100% (5233/5233), 4.99 MiB | 1.64 MiB/s, done.
Resolving deltas: 100% (3503/3503), done.


In [18]:
!cd mmdetection && python setup.py develop

Compiling mmdet/ops/nms/src/soft_nms_cpu.pyx because it changed.
[1/1] Cythonizing mmdet/ops/nms/src/soft_nms_cpu.pyx
/usr/local/lib/python3.6/dist-packages/setuptools/dist.py:474: UserWarning: Normalizing '1.0.rc0+0fe8df8' to '1.0rc0+0fe8df8'
  normalized_version,
running develop
running egg_info
creating mmdet.egg-info
writing mmdet.egg-info/PKG-INFO
writing dependency_links to mmdet.egg-info/dependency_links.txt
writing requirements to mmdet.egg-info/requires.txt
writing top-level names to mmdet.egg-info/top_level.txt
writing manifest file 'mmdet.egg-info/SOURCES.txt'
writing manifest file 'mmdet.egg-info/SOURCES.txt'
running build_ext
building 'mmdet.ops.nms.soft_nms_cpu' extension
creating build
creating build/temp.linux-x86_64-3.6
creating build/temp.linux-x86_64-3.6/mmdet
creating build/temp.linux-x86_64-3.6/mmdet/ops
creating build/temp.linux-x86_64-3.6/mmdet/ops/nms
creating build/temp.linux-x86_64-3.6/mmdet/ops/nms/src
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wa

# prepare coco dataset for an optional system pre-test

In [0]:
!wget http://images.cocodataset.org/annotations/annotations_trainval2017.zip

--2019-09-13 05:15:39--  http://images.cocodataset.org/annotations/annotations_trainval2017.zip
Resolving images.cocodataset.org (images.cocodataset.org)... 52.216.137.124
Connecting to images.cocodataset.org (images.cocodataset.org)|52.216.137.124|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 252907541 (241M) [application/zip]
Saving to: ‘annotations_trainval2017.zip’

annotations_trainva 100%[===================>] 241.19M  11.8MB/s    in 21s     

2019-09-13 05:16:01 (11.7 MB/s) - ‘annotations_trainval2017.zip’ saved [252907541/252907541]



In [0]:
!unzip  annotations_trainval2017.zip

Archive:  annotations_trainval2017.zip
  inflating: annotations/instances_train2017.json  
  inflating: annotations/instances_val2017.json  
  inflating: annotations/captions_train2017.json  
  inflating: annotations/captions_val2017.json  
  inflating: annotations/person_keypoints_train2017.json  
  inflating: annotations/person_keypoints_val2017.json  


In [0]:
!ls

annotations				      mmdetection
annotations_trainval2017.zip		      sample_data
annotations_trainval2017.zip.1		      val2017
faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth  val2017.zip


In [0]:
!mkdir  -p mmdetection/data/coco/

In [0]:
!cp -r ./annotations  mmdetection/data/coco

In [0]:
!ls mmdetection/data/coco/annotations/

captions_train2017.json   instances_val2017.json
captions_val2017.json	  person_keypoints_train2017.json
instances_train2017.json  person_keypoints_val2017.json


In [0]:
!wget http://images.cocodataset.org/zips/val2017.zip

--2019-09-13 05:17:10--  http://images.cocodataset.org/zips/val2017.zip
Resolving images.cocodataset.org (images.cocodataset.org)... 52.216.93.75
Connecting to images.cocodataset.org (images.cocodataset.org)|52.216.93.75|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 815585330 (778M) [application/zip]
Saving to: ‘val2017.zip’

val2017.zip         100%[===================>] 777.80M  16.6MB/s    in 49s     

2019-09-13 05:17:59 (15.9 MB/s) - ‘val2017.zip’ saved [815585330/815585330]



In [0]:
!ls

annotations				      mmdetection
annotations_trainval2017.zip		      sample_data
annotations_trainval2017.zip.1		      val2017
faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth  val2017.zip


In [0]:
!unzip val2017.zip

In [0]:
!ls

annotations				      mmdetection
annotations_trainval2017.zip		      sample_data
annotations_trainval2017.zip.1		      val2017
faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth  val2017.zip


In [0]:
!cp -r ./val2017  mmdetection/data/coco/

In [0]:
!wget  https://s3.ap-northeast-2.amazonaws.com/open-mmlab/mmdetection/models/faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth

--2019-09-13 05:18:48--  https://s3.ap-northeast-2.amazonaws.com/open-mmlab/mmdetection/models/faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth
Resolving s3.ap-northeast-2.amazonaws.com (s3.ap-northeast-2.amazonaws.com)... 52.219.60.4
Connecting to s3.ap-northeast-2.amazonaws.com (s3.ap-northeast-2.amazonaws.com)|52.219.60.4|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 167293562 (160M) [application/x-www-form-urlencoded]
Saving to: ‘faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth’

faster_rcnn_r50_fpn 100%[===================>] 159.54M  43.7MB/s    in 4.0s    

2019-09-13 05:18:53 (39.7 MB/s) - ‘faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth’ saved [167293562/167293562]



In [0]:
!wget  https://s3.ap-northeast-2.amazonaws.com/open-mmlab/mmdetection/models/cascade_rcnn_r50_fpn_1x_20190501-3b6211ab.pth

--2019-09-13 07:10:44--  https://s3.ap-northeast-2.amazonaws.com/open-mmlab/mmdetection/models/cascade_rcnn_r50_fpn_1x_20190501-3b6211ab.pth
Resolving s3.ap-northeast-2.amazonaws.com (s3.ap-northeast-2.amazonaws.com)... 52.219.60.16
Connecting to s3.ap-northeast-2.amazonaws.com (s3.ap-northeast-2.amazonaws.com)|52.219.60.16|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 277856190 (265M) [application/x-www-form-urlencoded]
Saving to: ‘cascade_rcnn_r50_fpn_1x_20190501-3b6211ab.pth’

cascade_rcnn_r50_fp 100%[===================>] 264.98M  35.4MB/s    in 7.9s    

2019-09-13 07:10:52 (33.7 MB/s) - ‘cascade_rcnn_r50_fpn_1x_20190501-3b6211ab.pth’ saved [277856190/277856190]



In [0]:
!mkdir -p mmdetection/checkpoints

In [0]:
!ls

annotations					    mmdetection
annotations_trainval2017.zip			    mydrive
cascade_mask_rcnn_r50_fpn_1x_20181123-88b170c9.pth  sample_data
cascade_rcnn_r50_fpn_1x_20190501-3b6211ab.pth	    val2017
faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth	    val2017.zip
gdrive


In [0]:
!cp cascade_rcnn_r50_fpn_1x_20190501-3b6211ab.pth mmdetection/checkpoints/

In [0]:
!cd mmdetection && python tools/test.py configs/faster_rcnn_r50_fpn_1x.py checkpoints/faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth --eval bbox --out results.pkl

loading annotations into memory...
Done (t=0.62s)
creating index...
index created!
terminal width is too small (0), please consider widen the terminal for better progressbar visualization
[          ] 82/5000, 2.7 task/s, elapsed: 30s, ETA:  1794sTraceback (most recent call last):
  File "tools/test.py", line 224, in <module>
    main()
  File "tools/test.py", line 184, in main
    outputs = single_gpu_test(model, data_loader, args.show)
  File "tools/test.py", line 26, in single_gpu_test
    result = model(return_loss=False, rescale=not show, **data)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 547, in __call__
    result = self.forward(*input, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/data_parallel.py", line 150, in forward
    return self.module(*inputs[0], **kwargs[0])
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 547, in __call__
    result = self.forward(*input, **kwargs)
  Fil

In [0]:
!ls -l annotations/

total 814872
-rw-rw-r-- 1 root root  91865115 Sep  1  2017 captions_train2017.json
-rw-rw-r-- 1 root root   3872473 Sep  1  2017 captions_val2017.json
-rw-rw-r-- 1 root root 469785474 Sep  1  2017 instances_train2017.json
-rw-rw-r-- 1 root root  19987840 Sep  1  2017 instances_val2017.json
-rw-rw-r-- 1 root root 238884731 Sep  1  2017 person_keypoints_train2017.json
-rw-rw-r-- 1 root root  10020657 Sep  1  2017 person_keypoints_val2017.json


In [0]:
from google.colab import files
files.download('./annotations/person_keypoints_val2017.json')


In [0]:
!ls

annotations				      mmdetection
annotations_trainval2017.zip		      sample_data
annotations_trainval2017.zip.1		      val2017
faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth  val2017.zip


# prepare  own dataset, and train a model, and generate a result, and submit

In [0]:
from google.colab import drive
drive.mount('/content/gdrive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/gdrive


In [0]:
!ls gdrive/'My Drive'/coco0828/

annotations  images


In [0]:
!pwd

/content


In [0]:
!ls

annotations				      gdrive	   val2017
annotations_trainval2017.zip		      mmdetection  val2017.zip
faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth  sample_data


In [0]:
!cd mmdetection/configs && git pull -p

remote: Enumerating objects: 7, done.
remote: Counting objects: 100% (7/7), done.
remote: Compressing objects: 100% (1/1), done.
remote: Total 4 (delta 3), reused 4 (delta 3), pack-reused 0
Unpacking objects: 100% (4/4), done.
From https://github.com/wangbingo/mmdetection
   d501f06..5581260  master     -> origin/master
Updating d501f06..5581260
Fast-forward
 configs/faster_rcnn_r50_fpn_1x.py | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)


In [0]:
!ls mmdetection

build	     data    docs	     mmdet	     requirements.txt  tools
checkpoints  demo    fabric2coco.py  mmdet.egg-info  setup.py
configs      docker  LICENSE	     README.md	     tests


In [0]:
!ln -s "/content/gdrive/My Drive" "/content/mydrive"

In [0]:
!ls -l /content/mydrive/coco0828/annotations/instances_train.json

-rw------- 1 root root 727622 Sep 12 19:32 /content/mydrive/coco0828/annotations/instances_train.json


In [0]:
!cd mmdetection/ && git pull -p

remote: Enumerating objects: 7, done.
remote: Counting objects: 100% (7/7), done.
remote: Compressing objects: 100% (1/1), done.
remote: Total 4 (delta 3), reused 4 (delta 3), pack-reused 0
Unpacking objects: 100% (4/4), done.
From https://github.com/wangbingo/mmdetection
   7e41596..f80aa0a  master     -> origin/master
Updating 7e41596..f80aa0a
Fast-forward
 configs/cascade_rcnn_r50_fpn_1x.py | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)


In [0]:
!cp mmdetection/tools/weights_num_chang.py  ./

In [0]:
!ls

annotations						 mmdetection
annotations_trainval2017.zip				 mydrive
cascade_mask_rcnn_r50_fpn_1x_20181123-88b170c9.pth	 sample_data
cascade_rcnn_r50_coco_pretrained_weights_classes_21.pth  val2017
cascade_rcnn_r50_fpn_1x_20190501-3b6211ab.pth		 val2017.zip
faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth		 weights_num_chang.py
gdrive


In [0]:
!python weights_num_chang.py

In [0]:
!cp cascade_rcnn_r50_coco_pretrained_weights_classes_21.pth  mmdetection/checkpoints/

In [0]:
!cp cascade_rcnn_r50_coco_pretrained_weights_classes_21.pth mydrive/checkpoints/

In [0]:
!cp *.pth mydrive/checkpoints/

# **test speed:** half size(1223*500) @ 3task/s, 2446*1000 @ 1.1task/s

In [0]:
!cd mmdetection && python tools/test.py configs/cascade_rcnn_r50_fpn_1x.py checkpoints/cascade_rcnn_r50_coco_pretrained_weights_classes_21.pth --eval bbox --out results.pkl

# **train params**: lr = 0.002, epochs = 3, training data = coco0828 (1139 imgs), 2446 * 1000, from scratch

In [0]:
!cd mmdetection && python tools/train.py configs/cascade_rcnn_r50_fpn_1x.py

2019-09-13 09:06:31,874 - INFO - Distributed training: False
2019-09-13 09:06:32,549 - INFO - load model from: torchvision://resnet50
2019-09-13 09:06:32,819 - WARNING - The model and loaded state dict do not match exactly

unexpected key in source state_dict: fc.weight, fc.bias

loading annotations into memory...
Done (t=0.02s)
creating index...
index created!
2019-09-13 09:06:35,775 - INFO - Start running, host: root@33f256921190, work_dir: /content/mmdetection/work_dirs/cascade_rcnn_r50_fpn_1x
2019-09-13 09:06:35,776 - INFO - workflow: [('train', 1)], max: 3 epochs
2019-09-13 09:07:44,192 - INFO - Epoch [1][50/570]	lr: 0.00080, eta: 0:37:50, time: 1.368, data_time: 0.013, memory: 2793, loss_rpn_cls: 0.2409, loss_rpn_bbox: 0.0268, s0.loss_cls: 0.3018, s0.acc: 95.7266, s0.loss_bbox: 0.0028, s1.loss_cls: 0.1652, s1.acc: 93.7910, s1.loss_bbox: 0.0004, s2.loss_cls: 0.1027, s2.acc: 90.4805, s2.loss_bbox: 0.0001, loss: 0.8406
2019-09-13 09:08:51,664 - INFO - Epoch [1][100/570]	lr: 0.00093,

In [0]:
!ls mmdetection/work_dirs/cascade_rcnn_r50_fpn_1x/ -l

total 1620420
-rw-r--r-- 1 root root       113 Sep 13 08:57 20190913_085713.log
-rw-r--r-- 1 root root       211 Sep 13 09:01 20190913_090155.log
-rw-r--r-- 1 root root     12154 Sep 13 09:44 20190913_090635.log
-rw-r--r-- 1 root root     12710 Sep 13 09:44 20190913_090635.log.json
-rw-r--r-- 1 root root 553084041 Sep 13 09:19 epoch_1.pth
-rw-r--r-- 1 root root 553084093 Sep 13 09:32 epoch_2.pth
-rw-r--r-- 1 root root 553084071 Sep 13 09:44 epoch_3.pth
lrwxrwxrwx 1 root root        11 Sep 13 09:44 latest.pth -> epoch_3.pth


# use epoch_3.pth to test at full size, speed is about 0.9~1.0 task/s

In [0]:
!cd mmdetection && python tools/test.py configs/cascade_rcnn_r50_fpn_1x.py work_dirs/cascade_rcnn_r50_fpn_1x/epoch_3.pth --eval bbox --out results.pkl

loading annotations into memory...
Done (t=0.83s)
creating index...
index created!
terminal width is too small (0), please consider widen the terminal for better progressbar visualization
[          ] 384/4774, 1.0 task/s, elapsed: 379s, ETA:  4334sTraceback (most recent call last):
  File "tools/test.py", line 224, in <module>
    main()
  File "tools/test.py", line 184, in main
    outputs = single_gpu_test(model, data_loader, args.show)
  File "tools/test.py", line 26, in single_gpu_test
    result = model(return_loss=False, rescale=not show, **data)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 547, in __call__
    result = self.forward(*input, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/data_parallel.py", line 150, in forward
    return self.module(*inputs[0], **kwargs[0])
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 547, in __call__
    result = self.forward(*input, **kwargs)
  F

In [0]:
!cd mmdetection/ && git pull -p

remote: Enumerating objects: 9, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (1/1), done.
remote: Total 5 (delta 4), reused 5 (delta 4), pack-reused 0
Unpacking objects: 100% (5/5), done.
From https://github.com/wangbingo/mmdetection
   f80aa0a..ded0c44  master     -> origin/master
Updating f80aa0a..ded0c44
Fast-forward
 .../faster_rcnn_r50_fpn_attention_0010_dcn_1x.py   | 30 +++++++++++++---------
 1 file changed, 18 insertions(+), 12 deletions(-)


## **train params**: num->21, train half size, test full size, 3 epochs, data from 1139 imgs

In [0]:
!cd mmdetection && python tools/train.py configs/empirical_attention/faster_rcnn_r50_fpn_attention_0010_dcn_1x.py

2019-09-13 10:31:35,942 - INFO - Distributed training: False
2019-09-13 10:31:36,453 - INFO - load model from: torchvision://resnet50
2019-09-13 10:31:36,732 - WARNING - The model and loaded state dict do not match exactly

unexpected key in source state_dict: fc.weight, fc.bias

missing keys in source state_dict: layer3.1.gen_attention_block.proj_conv.bias, layer3.5.gen_attention_block.value_conv.weight, layer4.1.gen_attention_block.key_conv.weight, layer4.2.gen_attention_block.proj_conv.weight, layer3.0.conv2_offset.bias, layer3.0.gen_attention_block.gamma, layer3.2.gen_attention_block.gamma, layer4.2.gen_attention_block.gamma, layer3.5.gen_attention_block.proj_conv.bias, layer4.0.gen_attention_block.gamma, layer3.2.conv2_offset.weight, layer3.4.gen_attention_block.value_conv.weight, layer4.1.conv2_offset.weight, layer4.1.gen_attention_block.proj_conv.bias, layer2.0.conv2_offset.bias, layer3.3.gen_attention_block.value_conv.weight, layer3.3.gen_attention_block.proj_conv.weight, layer

In [0]:
!cp   mmdetection/work_dirs/cascade_rcnn_r50_fpn_1x/epoch_3.pth  mydrive/checkpoints/cascade_rcnn_r50_fpn_1x_epoch_3.pth

In [0]:
!cd mmdetection/ && git pull -p

remote: Enumerating objects: 9, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (1/1), done.
remote: Total 5 (delta 4), reused 5 (delta 4), pack-reused 0
Unpacking objects: 100% (5/5), done.
From https://github.com/wangbingo/mmdetection
   ded0c44..50f919e  master     -> origin/master
Updating ded0c44..50f919e
Fast-forward
 .../faster_rcnn_r50_fpn_attention_0010_dcn_1x.py           | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)


# train params: num->21, train half size, test full size, 3 epochs, data from coco0818(4774 imgs)

In [0]:
!cd mmdetection && python tools/train.py configs/empirical_attention/faster_rcnn_r50_fpn_attention_0010_dcn_1x.py

2019-09-13 11:04:33,729 - INFO - Distributed training: False
2019-09-13 11:04:34,201 - INFO - load model from: torchvision://resnet50
2019-09-13 11:04:34,448 - WARNING - The model and loaded state dict do not match exactly

unexpected key in source state_dict: fc.weight, fc.bias

missing keys in source state_dict: layer3.4.gen_attention_block.proj_conv.bias, layer3.3.gen_attention_block.proj_conv.weight, layer3.0.gen_attention_block.key_conv.weight, layer3.3.gen_attention_block.key_conv.weight, layer4.0.gen_attention_block.proj_conv.weight, layer3.2.conv2_offset.bias, layer3.4.gen_attention_block.gamma, layer4.1.gen_attention_block.appr_bias, layer3.0.gen_attention_block.appr_bias, layer4.2.gen_attention_block.proj_conv.weight, layer2.0.conv2_offset.bias, layer4.2.conv2_offset.weight, layer3.5.gen_attention_block.gamma, layer4.2.gen_attention_block.gamma, layer3.1.gen_attention_block.proj_conv.weight, layer3.5.gen_attention_block.appr_bias, layer3.0.gen_attention_block.value_conv.weigh

In [0]:
!ls  mmdetection/work_dirs/faster_rcnn_r50_fpn_attention_0010_dcn_1x/ -l

total 1066576
-rw-r--r-- 1 root root      3256 Sep 13 10:48 20190913_103139.log
-rw-r--r-- 1 root root      3109 Sep 13 10:48 20190913_103139.log.json
-rw-r--r-- 1 root root     33304 Sep 13 14:04 20190913_110437.log
-rw-r--r-- 1 root root     33799 Sep 13 14:04 20190913_110437.log.json
-rw-r--r-- 1 root root 364026785 Sep 13 12:04 epoch_1.pth
-rw-r--r-- 1 root root 364026777 Sep 13 13:04 epoch_2.pth
-rw-r--r-- 1 root root 364026741 Sep 13 14:05 epoch_3.pth
lrwxrwxrwx 1 root root        11 Sep 13 14:05 latest.pth -> epoch_3.pth


In [0]:
!cp  mmdetection/work_dirs/faster_rcnn_r50_fpn_attention_0010_dcn_1x/epoch_3.pth  mydrive/checkpoints/faster_rcnn_r50_fpn_attention_0010_dcn_1x_epoch_3.pth

In [0]:
!cd mmdetection/ && git pull -p

remote: Enumerating objects: 9, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (1/1), done.
remote: Total 5 (delta 4), reused 5 (delta 4), pack-reused 0
Unpacking objects: 100% (5/5), done.
From https://github.com/wangbingo/mmdetection
   50f919e..cf28106  master     -> origin/master
Updating 50f919e..cf28106
Fast-forward
 .../empirical_attention/faster_rcnn_r50_fpn_attention_0010_dcn_1x.py    | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)


In [0]:
!cd mmdetection && python tools/test.py configs/empirical_attention/faster_rcnn_r50_fpn_attention_0010_dcn_1x.py work_dirs/faster_rcnn_r50_fpn_attention_0010_dcn_1x/epoch_3.pth --eval bbox --out results.pkl

loading annotations into memory...
Done (t=0.03s)
creating index...
index created!
terminal width is too small (0), please consider widen the terminal for better progressbar visualization
[>>>>>>>>>>] 1139/1139, 3.5 task/s, elapsed: 323s, ETA:     0s
writing results to results.pkl
Starting evaluate bbox
Loading and preparing results...
DONE (t=0.09s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=1.62s).
Accumulating evaluation results...
DONE (t=0.41s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.013
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.048
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.003
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.038
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.019
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.002
 Av

In [0]:
!cd mmdetection/ && git pull -p

remote: Enumerating objects: 6, done.
remote: Counting objects: 100% (6/6), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 4 (delta 2), reused 4 (delta 2), pack-reused 0
Unpacking objects: 100% (4/4), done.
From https://github.com/wangbingo/mmdetection
   cf28106..ce15835  master     -> origin/master
Updating cf28106..ce15835
Fast-forward
 tools/gen_result.py | 40 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 40 insertions(+)
 create mode 100644 tools/gen_result.py


In [0]:
!cd mmdetection/ && python tools/gen_result.py

19
10
19
3
10
19
20
17
3
4
15
19
4
15
17
10
19
3
17
4
12
13
14
4
15
17
10
19
20
17
19
3
4
12
13
3
3
4
10
19
3
4
1
2
4
7
14
15
17
3
4
15
17
15
4
15
17
1
4
17
3
10
19
20
19
3
3
19
19
3
10
10
19
12
3
1
4
7
12
13
15
17
4
15
17
3
4
4
10
17
19
20
4
7
15
17
17
3
10
19
14
3
10
19
17
3
4
14
19
17
1
4
15
15
17
14
19
4
15
17
3
14
19
3
4
17
19
4
15
17
15
17
4
15
3
4
10
19
3
10
19
20
19
17
12
4
17
3
3
4
4
15
17
3
12
3
4
17
3
10
19
1
3
4
3
4
14
15
17
19
3
4
4
15
17
3
4
3
4
3
3
3
4
1
4
1
3
4
15
3
4
4
15
12
13
14
4
7
15
17
19
4
15
17
4
15
4
4
15
4
15
17
4
7
15
17
1
4
15
17
17
3
4
17
12
13
3
4
19
12
13
1
3
4
15
15
17
3
4
15
19
4
7
15
17
19
1
3
4
12
13
15
3
4
19
17
10
19
4
17
3
4
15
17
3
10
19
20
3
14
19
1
3
4
1
4
12
13
3
4
17
4
10
17
19
3
4
3
4
17
3
4
15
1
3
4
15
1
4
12
13
1
3
4
15
1
3
4
12
13
12
13
4
15
17
3
10
1
4
15
3
4
15
12
14
3
4
17
3
4
3
4
3
4
1
4
15
17
1
3
4
15
17
10
19
20
4
12
13
17
17
4
15
17
1
3
4
3
10
19
20
10
17
10
19
20
4
15
17
3
4
17
17
3
3
4
10
19
19
3
4
1
3
4
1
3
4
15
1
3
4
10
19
19
3


In [0]:
!cd mmdetection/ && ls -l

total 5052
drwxr-xr-x  4 root root    4096 Sep 13 05:06 build
drwxr-xr-x  2 root root    4096 Sep 13 08:13 checkpoints
drwxr-xr-x 23 root root    4096 Sep 13 09:55 configs
drwxr-xr-x  3 root root    4096 Sep 13 05:16 data
drwxr-xr-x  2 root root    4096 Sep 13 05:06 demo
drwxr-xr-x  2 root root    4096 Sep 13 05:06 docker
drwxr-xr-x  2 root root    4096 Sep 13 05:06 docs
-rw-r--r--  1 root root    5191 Sep 13 05:06 fabric2coco.py
-rw-r--r--  1 root root   11400 Sep 13 05:06 LICENSE
drwxr-xr-x  9 root root    4096 Sep 13 05:19 mmdet
drwxr-xr-x  2 root root    4096 Sep 13 05:14 mmdet.egg-info
-rw-r--r--  1 root root    7152 Sep 13 05:06 README.md
-rw-r--r--  1 root root     100 Sep 13 05:06 requirements.txt
-rw-r--r--  1 root root 1464612 Sep 13 15:04 result_20190913144547.json
-rw-r--r--  1 root root 1613256 Sep 13 14:21 results.pkl
-rw-r--r--  1 root root 1999359 Sep 13 14:21 results.pkl.bbox.json
-rw-r--r--  1 root root    6472 Sep 13 05:06 setup.py
drwxr-xr-x  2 root root    4096 Sep

In [0]:
!cp mmdetection/result_20190913144547.json  mydrive/result/

# want to save the mmdetection dir to tar.gz, but the unzip env not work.a total lose.

In [0]:
!ls

annotations						 mmdetection
annotations_trainval2017.zip				 mydrive
cascade_mask_rcnn_r50_fpn_1x_20181123-88b170c9.pth	 sample_data
cascade_rcnn_r50_coco_pretrained_weights_classes_21.pth  val2017
cascade_rcnn_r50_fpn_1x_20190501-3b6211ab.pth		 val2017.zip
faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth		 weights_num_chang.py
gdrive


In [0]:
!cd mmdetection/ && rm  checkpoints/*

In [0]:
!cd mmdetection/work_dirs/   &&  rm -rf *

In [0]:
!ls  mmdetection/work_dirs/

In [0]:
!cd  mmdetection/data  &&  rm -rf   coco

In [0]:
!ls -l

total 2052048
drwxr-xr-x  2 root root      4096 Sep 13 05:16 annotations
-rw-r--r--  1 root root 252907541 Jul 10  2018 annotations_trainval2017.zip
-rw-r--r--  1 root root 309581110 Nov 26  2018 cascade_mask_rcnn_r50_fpn_1x_20181123-88b170c9.pth
-rw-r--r--  1 root root 277851208 Sep 13 08:12 cascade_rcnn_r50_coco_pretrained_weights_classes_21.pth
-rw-r--r--  1 root root 277856190 May  1 04:34 cascade_rcnn_r50_fpn_1x_20190501-3b6211ab.pth
-rw-r--r--  1 root root 167293562 Oct 12  2018 faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth
drwx------  3 root root      4096 Sep 13 05:24 gdrive
drwxr-xr-x 17 root root      4096 Sep 13 15:04 mmdetection
lrwxrwxrwx  1 root root        24 Sep 13 06:17 mydrive -> '/content/gdrive/My Drive'
drwxr-xr-x  1 root root      4096 Aug 27 16:17 sample_data
drwxrwxr-x  2 root root    180224 Aug 31  2017 val2017
-rw-r--r--  1 root root 815585330 Jul 11  2018 val2017.zip
-rw-r--r--  1 root root      1469 Sep 13 08:11 weights_num_chang.py


In [0]:
!rm  mmdetection.tar.gz

In [0]:
!tar   zcvf     mmdetection.tar.gz   mmdetection/

In [0]:
!ls  -l

total 2095472
drwxr-xr-x  2 root root      4096 Sep 13 05:16 annotations
-rw-r--r--  1 root root 252907541 Jul 10  2018 annotations_trainval2017.zip
-rw-r--r--  1 root root 309581110 Nov 26  2018 cascade_mask_rcnn_r50_fpn_1x_20181123-88b170c9.pth
-rw-r--r--  1 root root 277851208 Sep 13 08:12 cascade_rcnn_r50_coco_pretrained_weights_classes_21.pth
-rw-r--r--  1 root root 277856190 May  1 04:34 cascade_rcnn_r50_fpn_1x_20190501-3b6211ab.pth
-rw-r--r--  1 root root 167293562 Oct 12  2018 faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth
drwx------  3 root root      4096 Sep 13 05:24 gdrive
drwxr-xr-x 17 root root      4096 Sep 13 15:04 mmdetection
-rw-r--r--  1 root root  44465706 Sep 13 16:16 mmdetection.tar.gz
lrwxrwxrwx  1 root root        24 Sep 13 06:17 mydrive -> '/content/gdrive/My Drive'
drwxr-xr-x  1 root root      4096 Aug 27 16:17 sample_data
drwxrwxr-x  2 root root    180224 Aug 31  2017 val2017
-rw-r--r--  1 root root 815585330 Jul 11  2018 val2017.zip
-rw-r--r--  1 root root     

In [0]:
!cp mmdetection.tar.gz  mydrive/

In [0]:
!cd mmdetection/ && git pull -p

remote: Enumerating objects: 9, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (1/1), done.
remote: Total 5 (delta 4), reused 5 (delta 4), pack-reused 0
Unpacking objects: 100% (5/5), done.
From https://github.com/wangbingo/mmdetection
   ce15835..836dd80  master     -> origin/master
Updating ce15835..836dd80
Fast-forward
 .../faster_rcnn_r50_fpn_attention_0010_dcn_1x.py           | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)


In [0]:
!tar   zcvf     20190914-0025-mmdetection.tar.gz   mmdetection/

In [0]:
!cp 20190914-0025-mmdetection.tar.gz  mydrive/mmdetection.save/

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=email%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdocs.test%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.photos.readonly%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
!ln -s "/content/drive/My Drive" "/content/mydrive"

In [0]:
!cp  mydrive/mmdetection.save/20190914-0025-mmdetection.tar.gz  ./

In [0]:
!tar zxvf 20190914-0025-mmdetection.tar.gz

In [0]:
!rm 20190914-0025-mmdetection.tar.gz

In [10]:
!cd mmdetection/ && git pull -p

Already up to date.


#0914  train a attention model for 12 epochs

In [0]:
!cd mmdetection && python tools/train.py configs/empirical_attention/faster_rcnn_r50_fpn_attention_0010_dcn_1x.py

2019-09-13 17:48:16,245 - INFO - Distributed training: False
2019-09-13 17:48:16,654 - INFO - load model from: torchvision://resnet50
Downloading: "https://download.pytorch.org/models/resnet50-19c8e357.pth" to /root/.cache/torch/checkpoints/resnet50-19c8e357.pth
100% 97.8M/97.8M [00:03<00:00, 25.7MB/s]
2019-09-13 17:48:21,046 - WARNING - The model and loaded state dict do not match exactly

unexpected key in source state_dict: fc.weight, fc.bias

missing keys in source state_dict: layer4.0.gen_attention_block.value_conv.weight, layer4.0.conv2_offset.weight, layer4.2.gen_attention_block.key_conv.weight, layer3.1.gen_attention_block.proj_conv.bias, layer3.4.gen_attention_block.key_conv.weight, layer4.2.conv2_offset.bias, layer4.1.gen_attention_block.key_conv.weight, layer3.3.gen_attention_block.appr_bias, layer3.4.gen_attention_block.proj_conv.bias, layer3.4.gen_attention_block.proj_conv.weight, layer4.1.conv2_offset.bias, layer3.3.gen_attention_block.key_conv.weight, layer4.0.gen_attent

In [0]:
!cd mmdetection && python tools/train.py configs/cascade_rcnn_r50_fpn_1x.py